File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/compare/ModelCompare.kt
package org.kevoree.modeling.api.compare

import java.util.HashMap
import java.util.ArrayList
import org.kevoree.modeling.api.KMFContainer
import org.kevoree.modeling.api.trace.*;
import org.kevoree.modeling.api.KMFFactory

/**
 * Created by duke on 26/07/13.
 */

public class ModelCompare(val factory: KMFFactory) {

    public fun diff(origin: KMFContainer, target: KMFContainer): TraceSequence {
        return TraceSequence(factory).populate(internal_diff(origin, target, false, false));
    }

    public fun merge(origin: KMFContainer, target: KMFContainer): TraceSequence {
        return TraceSequence(factory).populate(internal_diff(origin, target, false, true));
    }

    public fun inter(origin: KMFContainer, target: KMFContainer): TraceSequence {
        return TraceSequence(factory).populate(internal_diff(origin, target, true, false));
    }

    private fun internal_diff(origin: KMFContainer, target: KMFContainer, inter: Boolean, merge: Boolean): List<ModelTrace> {
        val traces = ArrayList<ModelTrace>()
        val tracesRef = ArrayList<ModelTrace>()
        val objectsMap = HashMap <String, org.kevoree.modeling.api.KMFContainer>()
        traces.addAll(origin.createTraces(target, inter, merge, false, true))
        tracesRef.addAll(origin.createTraces(target, inter, merge, true, false))


        val visitor = object : org.kevoree.modeling.api.util.ModelVisitor() {
            override public fun visit(elem: org.kevoree.modeling.api.KMFContainer, refNameInParent: String, parent: org.kevoree.modeling.api.KMFContainer) {
                val childPath = elem.path();
                if (childPath != null) {
                    objectsMap.put(childPath, elem);
                } else {
                    throw Exception("Null child path " + elem);
                }
            }
        }
        origin.visit(visitor, true, true, false)

        val visitor2 = object : org.kevoree.modeling.api.util.ModelVisitor() {
            override public fun visit(elem: org.kevoree.modeling.api.KMFContainer, refNameInParent: String, parent: org.kevoree.modeling.api.KMFContainer) {
                val childPath = elem.path();
                if (objectsMap.containsKey(childPath)) {
                    if (inter) {
                        traces.add(ModelAddTrace(parent.path(), refNameInParent, elem.path(), elem.metaClassName()))
                    }
                    traces.addAll(objectsMap.get(childPath)!!.createTraces(elem, inter, merge, false, true))
                    tracesRef.addAll(objectsMap.get(childPath)!!.createTraces(elem, inter, merge, true, false))
                    objectsMap.remove(childPath) //drop from to process elements
                } else {
                    if (!inter) {
                        traces.add(ModelAddTrace(parent.path(), refNameInParent, elem.path(), elem.metaClassName()))
                        traces.addAll(elem.createTraces(elem, true, merge, false, true))
                        tracesRef.addAll(elem.createTraces(elem, true, merge, true, false))
                    }
                }
            }
        }
        target.visit(visitor2, true, true, false)
        traces.addAll(tracesRef); //references should be deleted before, deletion of elements
        if (!inter) {
            //if diff
            if (!merge) {
                for (diffChildKey in objectsMap.keySet()) {
                    val diffChild = objectsMap.get(diffChildKey)!!
                    val src = (if (diffChild.eContainer() != null) {
                        diffChild.eContainer()!!.path()
                    } else {
                        "null"
                    })
                    val refNameInParent = (if (diffChild.getRefInParent() != null) {
                        diffChild.getRefInParent()!!
                    } else {
                        "null"
                    })
                    traces.add(ModelRemoveTrace(src, refNameInParent, (diffChild as KMFContainer).path()))
                }
            }
        }
        return traces;
    }

}

File: cpp/org.kevoree.modeling.cpp.microframework/src/main/cpp/microframework/api/compare/ModelCompare.cpp
#include <microframework/api/compare/ModelCompare.h>
#include <microframework/api/utils/ModelVisitor.h>
#include <microframework/api/utils/Utils.h>
#include <microframework/api/compare/ModelCompareVisitors.h>
/**
 * Author: jedartois@gmail.com
 * Date: 24/10/13
 * Time: 18:36
 */


TraceSequence *ModelCompare::createSequence ()
{
	return new TraceSequence ();
}

TraceSequence* ModelCompare::diff (KMFContainer *origin, KMFContainer *target)
{
	return	createSequence ()->populate (internal_diff(origin, target, false, false));
}


TraceSequence *ModelCompare::merge (KMFContainer *origin, KMFContainer *target)
{
	return	createSequence ()->populate(internal_diff(origin, target, false, true));
}


TraceSequence *ModelCompare::inter (KMFContainer *origin, KMFContainer *target)
{
	return 	createSequence ()->populate (internal_diff(origin, target, true, false));
}


std::list < ModelTrace * >* ModelCompare::internal_diff (KMFContainer *origin,KMFContainer *target,bool inter, bool merge)
{
	list < ModelTrace * > *traces ;
	list < ModelTrace * > *tracesRef;
	std::map<string,KMFContainer*> values;

	traces = origin->createTraces (target, inter, merge, false, true);
	tracesRef = origin->createTraces (target, inter, merge, true, false);

	// visitors 
	ModelCompareVisitorFiller *filler = new ModelCompareVisitorFiller(&values);
	origin->visit(filler, true, true, false);
	delete filler;


	ModelCompareVisitorCreateTraces *visitorTraces= new ModelCompareVisitorCreateTraces(&values,inter,merge,traces,tracesRef);
	target->visit(visitorTraces, true, true, false);
	delete visitorTraces;

	if(!inter){
		//if diff
		if(!merge)
		{
			for ( std::map<string,KMFContainer*>::const_iterator it = values.begin();  it != values.end(); ++it) {

				KMFContainer *diffChild;
				string src;
				string refNameInParent="";
				diffChild =it->second;

				if( diffChild-> eContainer())
				{
					src =   diffChild-> eContainer()->path();
				}
				refNameInParent = diffChild-> getRefInParent();
				ModelRemoveTrace *modelremovetrace= new ModelRemoveTrace(src,refNameInParent,diffChild->path());
				traces->push_back(modelremovetrace);
			}
		}
	}
	std::copy(tracesRef->begin(), tracesRef->end(), std::back_insert_iterator<std::list<ModelTrace*> >(*traces));
	values.clear();
	delete tracesRef;

	return traces;
}
